## Film Strategy Study
This notebook is being used to explore the data available through the im.db database.

In [1]:
import pandas as pd
import sqlite3

In [2]:
!pwd

/c/Users/NM/Documents/Flatiron/phase_1/Film-Strategy/Study


In [3]:
conn = sqlite3.connect('../zippedData/im.db')

**Let's look at the db schema.**

In [4]:
query = """
SELECT *
FROM sqlite_master
"""
pd.read_sql_query(query, conn)

type           name       tbl_name  rootpage  \
0  table   movie_basics   movie_basics         2   
1  table      directors      directors         3   
2  table      known_for      known_for         4   
3  table     movie_akas     movie_akas         5   
4  table  movie_ratings  movie_ratings         6   
5  table        persons        persons         7   
6  table     principals     principals         8   
7  table        writers        writers         9   

                                                 sql  
0  CREATE TABLE "movie_basics" (\n"movie_id" TEXT...  
1  CREATE TABLE "directors" (\n"movie_id" TEXT,\n...  
2  CREATE TABLE "known_for" (\n"person_id" TEXT,\...  
3  CREATE TABLE "movie_akas" (\n"movie_id" TEXT,\...  
4  CREATE TABLE "movie_ratings" (\n"movie_id" TEX...  
5  CREATE TABLE "persons" (\n"person_id" TEXT,\n ...  
6  CREATE TABLE "principals" (\n"movie_id" TEXT,\...  
7  CREATE TABLE "writers" (\n"movie_id" TEXT,\n  ...

**Examining the movie_basics table**

In [5]:
query = """
SELECT *
FROM movie_basics
GROUP BY movie_id
-- LIMIT 5
"""
pd.read_sql_query(query,conn)

movie_id                                primary_title  \
0       tt0063540                                    Sunghursh   
1       tt0066787              One Day Before the Rainy Season   
2       tt0069049                   The Other Side of the Wind   
3       tt0069204                              Sabse Bada Sukh   
4       tt0100275                     The Wandering Soap Opera   
...           ...                                          ...   
146139  tt9916538                          Kuambil Lagi Hatiku   
146140  tt9916622  Rodolpho Teóphilo - O Legado de um Pioneiro   
146141  tt9916706                              Dankyavar Danka   
146142  tt9916730                                       6 Gunn   
146143  tt9916754               Chico Albuquerque - Revelações   

                                     original_title  start_year  \
0                                         Sunghursh        2013   
1                                   Ashad Ka Ek Din        2019   
2                        The Other Side of the Wind        2018   
3                                   Sabse Bada Sukh        2018   
4                             La Telenovela Errante        2017   
...                                             ...         ...   
146139                          Kuambil Lagi Hatiku        2019   
146140  Rodolpho Teóphilo - O Legado de um Pioneiro        2015   
146141                              Dankyavar Danka        2013   
146142                                       6 Gunn        2017   
146143               Chico Albuquerque - Revelações        2013   

        runtime_minutes                genres  
0                 175.0    Action,Crime,Drama  
1                 114.0       Biography,Drama  
2                 122.0                 Drama  
3                   NaN          Comedy,Drama  
4                  80.0  Comedy,Drama,Fantasy  
...                 ...                   ...  
146139            123.0                 Drama  
146140              NaN           Documentary  
146141              NaN                Comedy  
146142            116.0                  None  
146143              NaN           Documentary  

[146144 rows x 6 columns]

**movie_id is probably the value needed to connect to other tables such as director, writer, etc.**
**Let's look at the director table**

In [6]:
query = """
SELECT *
FROM directors
LIMIT 5
"""
pd.read_sql_query(query, conn)

movie_id  person_id
0  tt0285252  nm0899854
1  tt0462036  nm1940585
2  tt0835418  nm0151540
3  tt0835418  nm0151540
4  tt0878654  nm0089502

**Presumably, for the movie represented by movie_id, the person_id points to detail about the director. Implication is that person_id is probably to be found in the table called persons.**

In [7]:
query = """
SELECT *
FROM persons
LIMIT 5
"""
pd.read_sql_query(query, conn)

person_id       primary_name birth_year death_year  \
0  nm0061671  Mary Ellen Bauder       None       None   
1  nm0061865       Joseph Bauer       None       None   
2  nm0062070         Bruce Baum       None       None   
3  nm0062195       Axel Baumann       None       None   
4  nm0062798        Pete Baxter       None       None   

                                 primary_profession  
0         miscellaneous,production_manager,producer  
1        composer,music_department,sound_department  
2                        miscellaneous,actor,writer  
3  camera_department,cinematographer,art_department  
4  production_designer,art_department,set_decorator

**What does movie_ratings table contain?**

In [8]:
query = """

SELECT *
FROM movie_ratings
LIMIT 5
"""
pd.read_sql_query(query, conn)

movie_id  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

**So, movie_ratings table provides the average rating across the given number of votes for the movie represented by movie_id.**

In [ ]:
Used to randomly look at movie records by manually providing a movie_id.

In [9]:
query = """
SELECT *
FROM movie_basics
WHERE movie_id = 'tt10356526'
LIMIT 5
"""
pd.read_sql_query(query,conn)

movie_id     primary_title    original_title  start_year  \
0  tt10356526  Laiye Je Yaarian  Laiye Je Yaarian        2019   

   runtime_minutes   genres  
0            117.0  Romance

**Let's join the movie name and info with the director and ratings.**

In [10]:
query = """
SELECT DISTINCT *
FROM movie_basics
LEFT JOIN directors
ON movie_basics.movie_id = directors.movie_id
LEFT JOIN persons
ON directors.person_id = persons.person_id
LEFT JOIN movie_ratings
ON movie_basics.movie_id = movie_ratings.movie_id
LIMIT 10
"""
pd.read_sql_query(query, conn)

movie_id                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   
5  tt0100275         The Wandering Soap Opera       La Telenovela Errante   
6  tt0111414                      A Thin Life                 A Thin Life   
7  tt0112502                          Bigfoot                     Bigfoot   
8  tt0137204                  Joe Finds Grace             Joe Finds Grace   
9  tt0139613                       O Silêncio                  O Silêncio   

   start_year  runtime_minutes                      genres   movie_id  \
0        2013            175.0          Action,Crime,Drama  tt0063540   
1        2019            114.0             Biography,Drama  tt0066787   
2        2018            122.0                       Drama  tt0069049   
3        2018              NaN                Comedy,Drama  tt0069204   
4        2017             80.0        Comedy,Drama,Fantasy  tt0100275   
5        2017             80.0        Comedy,Drama,Fantasy  tt0100275   
6        2018             75.0                      Comedy  tt0111414   
7        2017              NaN             Horror,Thriller  tt0112502   
8        2017             83.0  Adventure,Animation,Comedy  tt0137204   
9        2012              NaN         Documentary,History  tt0139613   

   person_id  person_id               primary_name  birth_year  death_year  \
0  nm0712540  nm0712540        Harnam Singh Rawail      1921.0      2004.0   
1  nm0002411  nm0002411                  Mani Kaul      1944.0      2011.0   
2  nm0000080  nm0000080               Orson Welles      1915.0      1985.0   
3  nm0611531  nm0611531       Hrishikesh Mukherjee      1922.0      2006.0   
4  nm0749914  nm0749914                 Raoul Ruiz      1941.0      2011.0   
5  nm0765384  nm0765384          Valeria Sarmiento      1948.0         NaN   
6  nm0398271  nm0398271               Frank Howson      1952.0         NaN   
7  nm6883878  nm6883878                   Mc Jones         NaN         NaN   
8  nm0365480  nm0365480           Anthony Harrison      1961.0         NaN   
9  nm0023406  nm0023406  José Manuel Alves Pereira         NaN         NaN   

         primary_profession   movie_id  averagerating  numvotes  
0  director,writer,producer  tt0063540            7.0      77.0  
1     director,writer,actor  tt0066787            7.2      43.0  
2     actor,director,writer  tt0069049            6.9    4517.0  
3    director,editor,writer  tt0069204            6.1      13.0  
4  director,writer,producer  tt0100275            6.5     119.0  
5    editor,director,writer  tt0100275            6.5     119.0  
6     actor,writer,producer       None            NaN       NaN  
7            actor,director  tt0112502            4.1      32.0  
8     actor,writer,producer  tt0137204            8.1     263.0  
9  editor,director,producer       None            NaN       NaN

**Trying to be a little more judicious with the number of columns being selected.**
**Many of the top scoring movies seem to have relatively few votes.**
**Lets look at movies with at least 500 votes and limit display to first 100 rows.**

In [15]:
query = """
SELECT DISTINCT mb.primary_title AS "Movie Name", mb.start_year AS "Release Year", p.primary_name AS "Director", mr.averagerating AS "Avg Rating", mr.numvotes AS "Number of Votes"

FROM movie_basics mb
    
    LEFT JOIN directors d
        ON mb.movie_id = d.movie_id
    LEFT JOIN persons p
        ON d.person_id = p.person_id
    LEFT JOIN movie_ratings mr
        ON mb.movie_id = mr.movie_id

WHERE mr.numvotes > 500

ORDER BY mr.averagerating DESC

LIMIT 100
"""
pd.read_sql_query(query, conn)

Movie Name  Release Year              Director  \
0   Once Upon a Time ... in Hollywood          2019     Quentin Tarantino   
1                           Eghantham          2018        Arsel Arumugam   
2           Yeh Suhaagraat Impossible          2019        Abhinav Thakur   
3                 Ananthu V/S Nusrath          2018    Sudheer Shanbhogue   
4                 Ekvtime: Man of God          2018  Nikoloz Khomasuridze   
..                                ...           ...                   ...   
95                        Be Here Now          2015        Lilibet Foster   
96           A Place in the Caribbean          2017   Juan Carlos Fanconi   
97                             Sachin          2017         James Erskine   
98                          Farmlands          2018       Lauren Southern   
99                         Coriolanus          2014       Tim Van Someren   

    Avg Rating  Number of Votes  
0          9.7             5600  
1          9.7              639  
2          9.6              624  
3          9.6              808  
4          9.6             2604  
..         ...              ...  
95         8.7             2746  
96         8.7             1250  
97         8.7             7122  
98         8.7              791  
99         8.7             1347  

[100 rows x 5 columns]

**Capturing the query in a data frame without limit on number of rows**

In [22]:
query = """
SELECT DISTINCT mb.primary_title AS "Movie Name", mb.start_year AS "Release Year", p.primary_name AS "Director", mr.averagerating AS "Avg Rating", mr.numvotes AS "Number of Votes"

FROM movie_basics mb
    
    LEFT JOIN directors d
        ON mb.movie_id = d.movie_id
    LEFT JOIN persons p
        ON d.person_id = p.person_id
    LEFT JOIN movie_ratings mr
        ON mb.movie_id = mr.movie_id

WHERE mr.numvotes > 500

ORDER BY mr.averagerating DESC

"""
imdb_df = pd.read_sql_query(query, conn)
len(imdb_df)

15595

**Out of the original 146,144 rows in the movie_basics table, only 15,595 rows meet this criteria.**

In [17]:
imdb_df.head(100)

Movie Name  Release Year              Director  \
0   Once Upon a Time ... in Hollywood          2019     Quentin Tarantino   
1                           Eghantham          2018        Arsel Arumugam   
2           Yeh Suhaagraat Impossible          2019        Abhinav Thakur   
3                 Ananthu V/S Nusrath          2018    Sudheer Shanbhogue   
4                 Ekvtime: Man of God          2018  Nikoloz Khomasuridze   
..                                ...           ...                   ...   
95                        Be Here Now          2015        Lilibet Foster   
96           A Place in the Caribbean          2017   Juan Carlos Fanconi   
97                             Sachin          2017         James Erskine   
98                          Farmlands          2018       Lauren Southern   
99                         Coriolanus          2014       Tim Van Someren   

    Avg Rating  Number of Votes  
0          9.7             5600  
1          9.7              639  
2          9.6              624  
3          9.6              808  
4          9.6             2604  
..         ...              ...  
95         8.7             2746  
96         8.7             1250  
97         8.7             7122  
98         8.7              791  
99         8.7             1347  

[100 rows x 5 columns]

In [19]:
imdb_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15595 entries, 0 to 15594
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Movie Name       15595 non-null  object 
 1   Release Year     15595 non-null  int64  
 2   Director         15588 non-null  object 
 3   Avg Rating       15595 non-null  float64
 4   Number of Votes  15595 non-null  int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 609.3+ KB


In [28]:
imdb_df.loc[imdb_df['Director'] == 'Quentin Tarantino']['Movie Name']

0        Once Upon a Time ... in Hollywood
74      Kill Bill: The Whole Bloody Affair
217                       Django Unchained
1002                     The Hateful Eight
Name: Movie Name, dtype: object

In [31]:
imdb_df['Director'].isnull().sum()

7

In [42]:
query = """
SELECT DISTINCT mb.primary_title AS "Movie Name", mb.start_year AS "Release Year", p.primary_name AS "Director", mr.averagerating AS "Avg Rating", mr.numvotes AS "Number of Votes"

FROM movie_basics mb
    
    LEFT JOIN directors d
        ON mb.movie_id = d.movie_id
    LEFT JOIN persons p
        ON d.person_id = p.person_id
    LEFT JOIN movie_ratings mr
        ON mb.movie_id = mr.movie_id

WHERE mr.numvotes > 500 AND p.primary_name IS Null

ORDER BY mr.averagerating DESC

"""
pd.read_sql_query(query, conn)

Movie Name  Release Year Director  Avg Rating  \
0                 On vam ne Dimon          2017     None         9.2   
1                  Burning Love 2          2014     None         8.0   
2                    Free to Play          2014     None         8.0   
3  Digimon Adventure Tri. 4: Loss          2017     None         7.7   
4   Horoob Etirari: Forced Escape          2017     None         6.0   
5                         Ecstasy          2011     None         4.3   
6           Rise of the Black Bat          2012     None         1.2   

   Number of Votes  
0             2721  
1              531  
2            12704  
3              521  
4              802  
5              721  
6              616